src/main/resources/develocity-injection.init.gradle

```diff
@@ -197,41 +197,47 @@ void enableDevelocityInjection() {
                         }
                     }
 
-                    if (develocityUrl && develocityEnforceUrl) {
-                        logger.lifecycle("Enforcing Develocity: $develocityUrl, allowUntrustedServer: $develocityAllowUntrustedServer, captureFileFingerprints: $develocityCaptureFileFingerprints")
-                    }
-
                     pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
-                        // Only execute if develocity plugin isn't applied.
+                        // Only execute if Develocity plugin isn't applied.
                         if (gradle.rootProject.extensions.findByName("develocity")) return
-                        afterEvaluate {
-                            if (develocityUrl && develocityEnforceUrl) {
-                                buildScan.server = develocityUrl
-                                buildScan.allowUntrustedServer = develocityAllowUntrustedServer
-                            }
-                        }
-
                         if (buildScanTermsOfUseUrl && buildScanTermsOfUseAgree) {
                             buildScan.termsOfServiceUrl = buildScanTermsOfUseUrl
                             buildScan.termsOfServiceAgree = buildScanTermsOfUseAgree
                         }
                     }
 
                     pluginManager.withPlugin(DEVELOCITY_PLUGIN_ID) {
-                        afterEvaluate {
-                            if (develocityUrl && develocityEnforceUrl) {
-                                develocity.server = develocityUrl
-                                develocity.allowUntrustedServer = develocityAllowUntrustedServer
-                            }
-                        }
-
                         if (buildScanTermsOfUseUrl && buildScanTermsOfUseAgree) {
                             develocity.buildScan.termsOfUseUrl = buildScanTermsOfUseUrl
                             develocity.buildScan.termsOfUseAgree = buildScanTermsOfUseAgree
                         }
                     }
                 }
 
+                if (develocityUrl && develocityEnforceUrl) {
+                    logger.lifecycle("Enforcing Develocity: $develocityUrl, allowUntrustedServer: $develocityAllowUntrustedServer")
+                }
+
+                pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
+                    afterEvaluate {
+                        // Only execute if Develocity plugin isn't applied.
+                        if (pluginManager.hasPlugin(DEVELOCITY_PLUGIN_ID)) return
```

Comment:
@bigdaz on line 202, the existing logic checks for the availability of the Develocity plugin by checking for the existence of the `develocity` extension on the root project.

https://github.com/gradle/develocity-ci-injection/blob/5c039f7b8d3b99c7559054c48c71940fb43a7fcb/src/main/resources/develocity-injection.init.gradle#L202

I tried the same thing here, but that didn't work in all cases. The `com.gradle:build-scan-plugin:3.17.+` _**also**_ applies the `develocity` extension which would mean this block wouldn't be executed and the URL wouldn't be enforced. 

Is there a reason you check if the Develocity plugin is applied by checking for the `develocity` extension instead of checking if the plugin is applied directly (i.e., `pluginManager.hasPlugin(DEVELOCITY_PLUGIN_ID)`)?

Reply:
> I think that newer build scan plugins will also apply the Develocity plugin. So in this case the second `withPlugin(DEVELOCITY_PLUGIN_ID)` block will be applied.

This doesn't appear to be the case. In [this build scan](https://ge.solutions-team.gradle.com/s/edmgjrvjkgovm/plugins#com.gradle.build-scan-0) I applied `com.gradle.build-scan:3.18.1` and the Develocity plugin wasn't also applied. However, `com.gradle.build-scan:3.18.1` _does_ create a `develocity` extension. 

With the previous logic, this creates a scenario where neither block is called when only `com.gradle.build-scan:3.17+` is applied:

```groovy
pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
  if (gradle.rootProject.extensions.findByName("develocity")) return
  // develocity extension exists, never makes it here ❌
}

pluginManager.withPlugin(DEVELOCITY_PLUGIN_ID) {
  // develocity plugin not applied, never makes it here ❌
}
```

If we check for the existence of the plugin, then in the same scenario, the first block will be called:

```groovy
pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
  if (pluginManager.hasPlugin(DEVELOCITY_PLUGIN_ID)) return
  // develocity plugin not applied, this logic executes ✅
}

pluginManager.withPlugin(DEVELOCITY_PLUGIN_ID) {
  // develocity plugin not applied, never makes it here ❌
}
```

> If there's a `develocity` extension, configure it directly
> If there's no `develocity` extension, configure the `buildScan` extension.

This is what we do for Gradle 6+ so this makes sense to me. Unfortunately, it seems neither the previous logic nor what I have actually does this.

A somewhat crude way to get this to work would be:

```groovy
pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
  if (gradle.rootProject.extensions.findByName("develocity")) {
    // configure the 'develocity' extension
  } else {
    // configure the 'buildScan' extension
  }
}

pluginManager.withPlugin(DEVELOCITY_PLUGIN_ID) {
  // configure the 'develocity' extension
}
```

This should cover all cases, but maybe you know of a better way?

Reply:
> > I think that newer build scan plugins will also apply the Develocity plugin. So in this case the second `withPlugin(DEVELOCITY_PLUGIN_ID)` block will be applied.
> 
> This doesn't appear to be the case. In [this build scan](https://ge.solutions-team.gradle.com/s/edmgjrvjkgovm/plugins#com.gradle.build-scan-0) I applied `com.gradle.build-scan:3.18.1` and the Develocity plugin wasn't also applied. However, `com.gradle.build-scan:3.18.1` _does_ create a `develocity` extension.

Ah right, I think maybe I got this back-to-front. Could it be that applying the Develocity plugin automatically applies the Build Scan plugin? 

> 
> With the previous logic, this creates a scenario where neither block is called when only `com.gradle.build-scan:3.17+` is applied:
> 
> ```groovy
> pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
>   if (gradle.rootProject.extensions.findByName("develocity")) return
>   // develocity extension exists, never makes it here ❌
> }
> 
> pluginManager.withPlugin(DEVELOCITY_PLUGIN_ID) {
>   // develocity plugin not applied, never makes it here ❌
> }
> ```

Good catch. I guess this functionality wasn't tested properly with the Build Scan plugin applied. 

> 
> > If there's a `develocity` extension, configure it directly
> > If there's no `develocity` extension, configure the `buildScan` extension.
> 
> This is what we do for Gradle 6+ so this makes sense to me. Unfortunately, it seems neither the previous logic nor what I have actually does this.
> 
> A somewhat crude way to get this to work would be:
> 
> ```groovy
> pluginManager.withPlugin(BUILD_SCAN_PLUGIN_ID) {
>   if (gradle.rootProject.extensions.findByName("develocity")) {
>     // configure the 'develocity' extension
>   } else {
>     // configure the 'buildScan' extension
>   }
> }
> 
> pluginManager.withPlugin(DEVELOCITY_PLUGIN_ID) {
>   // configure the 'develocity' extension
> }
> ```
> 
> This should cover all cases, but maybe you know of a better way?

The problem with this logic is that I think it will result in the develocity extension being configured twice when the Build Scan plugin is applied. Since we are reacting via `withPlugin()`, I think your `hasPlugin()` check is probably the best way to avoid this.

In your PR the `hasPlugin(DEVELOCITY_PLUGIN_ID)` check is inside `afterEvaluate`. Is this required?

Reply:
> Could it be that applying the Develocity plugin automatically applies the Build Scan plugin?

That does appear to be the case. In Gradle 5.x, applying the Develocity plugin [also applies the Build Scan plugin](https://ge.solutions-team.gradle.com/s/nt2ev22ful6ha/plugins#com.gradle.build-scan-0). In Gradle 6+, [the Gradle Enterprise plugin is applied](https://ge.solutions-team.gradle.com/s/kmigcvpwhgck4/plugins#com.gradle.enterprise-10).

> Good catch. I guess this functionality wasn't tested properly with the Build Scan plugin applied.

I do think there is a gap in the tested versions. When we test a "`legacy`" plugin, we only test up to version 3.16.2, which doesn't create this scenario where you have both the legacy and new extensions. I think we additionally need to test the legacy plugins at the same version as the Develocity plugin, keeping in mind that eventually we'll stop distributing the legacy plugins.

https://github.com/gradle/develocity-ci-injection/blob/b441b7540a29b749075dda9314cb774e7459474c/src/test/groovy/com/gradle/BaseInitScriptTest.groovy#L174

> The problem with this logic is that I think it will result in the develocity extension being configured twice when the Build Scan plugin is applied. Since we are reacting via `withPlugin()`, I think your `hasPlugin()` check is probably the best way to avoid this.

Good point.

> In your PR the `hasPlugin(DEVELOCITY_PLUGIN_ID)` check is inside `afterEvaluate`. Is this required?

Probably not. It's probably better to do it outside the `afterEvaluate` so the callback isn't even registered in the first place.

Reply:
Resolving this conversation as we seem to have found a solution that works in all cases.